client/allocrunner/taskrunner/template/template.go

```diff
@@ -494,6 +508,74 @@ func (tm *TaskTemplateManager) onTemplateRendered(handledRenders map[string]time
 		}
 	}
 
+	// process script execution concurrently
+	var wg sync.WaitGroup
+	for _, script := range scripts {
+		wg.Add(1)
+		go tm.processScript(script, &wg)
+	}
+	wg.Wait()
+}
+
+// failOnError is a helper function that produces a TaskKilling event and emits
+// a message
+func (tm *TaskTemplateManager) failOnError(script *structs.ChangeScript, msg string) {
+	if script.FailOnError {
+		tm.config.Lifecycle.Kill(context.Background(),
+			structs.NewTaskEvent(structs.TaskKilling).
+				SetFailsTask().
+				SetDisplayMessage(msg))
+	}
+}
+
+// processScript is used for executing change_mode script and handling errors
+func (tm *TaskTemplateManager) processScript(script *structs.ChangeScript, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if tm.handle == nil {
+		failure_msg := fmt.Sprintf(
+			"Template failed to run script %v with arguments %v because task driver doesn't support the exec operation",
+			script.Command,
+			script.Args,
+		)
+		tm.config.Events.EmitEvent(structs.NewTaskEvent(structs.TaskHookFailed).
+			SetDisplayMessage(failure_msg))
+		tm.failOnError(script, failure_msg+" Task is being killed.")
+		return
+	}
+	_, exitCode, err := tm.handle.Exec(script.Timeout, script.Command, script.Args)
+	if err != nil {
+		failure_msg := fmt.Sprintf(
+			"Template failed to run script %v with arguments %v on change: %v Exit code: %v",
+			script.Command,
+			script.Args,
+			err,
+			exitCode,
+		)
+		structs.NewTaskEvent(structs.TaskHookFailed).SetDisplayMessage(failure_msg)
```

Comment:
I think this (and the next) events are not being emitted 😅

Reply:
oh wow what a fail. thanks for catching this Luiz.